Introduce variables in profile paths

mgerstner opened the pull request:
This is a first proposal to introduce variables in profile paths. Once we agree on the basic syntax, config file and implementation I'll add some additional commits to document and test the new feature.

mgerstner committed:
chkstat: fix some typos and whitespace

mgerstner committed:
chkstat: prepare reuse of config file locations

mgerstner committed:
chkstat: implement support for variables in profile paths

In situations like in bsc#1171164 (libexec path change) or PR#104
(moving binaries from /sbin to /usr/sbin) or simply in case of
architectural differences (/lib64 vs. /lib) it is currently necessary to
create duplicate or even triplicate profile path entries. This makes
maintenance of the profiles harder, the possiblity to introduce mistakes
or inconsistencies is higher and checking correctness of the profiles
(whitelisting checkers) becomes harder.

To address this issue this commit introduces a new variable syntax in
profile path specifications. A new configuration file "variables.conf"
contains variable definitions. Each variable can have one or more path
values. A variable in profile paths uses the "%{variable}" syntax. '%'
is used instead of '$', because '$' is already treated as a comment
character for historical reasons.

Each occurence of a "%{variable}" in a profile path will be expanded to
the possible variable's values, thereby creating possibly multiple
profile entries from the same profile line. Therefore a context like
this:

```
lib_paths = /usr/lib /usr/lib64

%{lib_paths}/squid/pinger root:squid 0750
```

will cause the respective pinger permissions to be applied to occurences
in both /usr/lib/squid/pinger and /usr/lib64/squid/pinger.

maltek commented on src/chkstat.cpp:
```diff
@@ -612,8 +639,83 @@ bool Chkstat::parseProfile(const std::string &path)
+    // path components i.e. something like %{myvar}stuff/suffix is not
+    // allowed, only %{myvar}/suffix.
+    //
+    // multiple variable components in the same path are supported
```
Implicitly this parsing means we don't support slashes in variable *names*. That's perfectly fine. I'm just wondering if we should make this more explicit.

mgerstner replied:
I think no special characters should be allowed in variable names. alphanumerical only. plus underscore maybe. I can add checking logic for that.

maltek replied:
Also makes sense :)

maltek commented on src/chkstat.cpp:
```diff
@@ -612,8 +639,83 @@ bool Chkstat::parseProfile(const std::string &path)
+            {
+                for (const auto &var_value: it->second)
+                {
+                    std::string sep = var_value[0] == '/' ? "" : "/";
```
Can `var_value` ever be empty here? Maybe we can enforce all variable values starting with a slash when reading them in?

mgerstner replied:
Since the parser currently splits words at whitespace boundaries for the variable values no value should ever be empty. Even then we would only access the string terminator here.

I thought about requiring them starting with a slash, too. I fear, however, that this could limit the usability of variables or at least make them strange to use if you ever happen to want it to contain only a single path component.

mgerstner replied:
Maybe I should normalize the variables by removing leading slashes and also duplicate ones and trailing ones. How about that?

maltek replied:
Yes, that makes sense.

maltek reviewed (commented):
Looking good overall.

The commit message still talks about "${variable}" (with $) in the third paragraph.

Some tests would be nice, including for the edge cases I brought up.

mgerstner commented on src/utility.h:
```diff
@@ -44,6 +44,42 @@ class SwitchArgRW :
+ *  This is just a small wrapper to fix this situation.
+ **/
+template <typename T>
+class SaneValueArg :
```
Yeah I don't know ... also sounds kind of strange, since it isn't really const, it just supports const usage.

maltek replied:
True. I don't have any other good ideas either. ('Constable' has a very different meaning.)

mgerstner replied:
I'd say we leave it as is

mgerstner commented:
The additional comit contains two additions for limiting supported variable characters and for normalizing variable value paths. I also adjusted the bad commit comment.

mgerstner committed:
man pages: add documentation about variables, update copyrights

mgerstner committed:
regtests: split up base types and actual test implementation

The single python file started becoming pretty long so lets at least
split up base infrastructure and test case implementations.

mgerstner committed:
tests: avoid redundant specification of test names by using class names

mgerstner committed:
tests: implement basic tests for new the new variable feature

mgerstner commented:
I added commits for the documentation and the test coverage.

maltek reviewed (approved):
LGTM
